### PR TITLE
Admission: support edge apps autonomy when offline

### LINKED ISSUE #2385: pod would restart when long time offline
**What happened and what you expected to happen**:

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

**Environment**:
<details><summary>Kubernets Version</summary>

```console
$ kubectl --version
# paste output here
```

</details>

<details><summary>CloudCore Version</summary>

```console
$ cloudcore --version
# paste output here
```

</details>

<details><summary>EdgeCore Version</summary>

```
$ edgecore --version
# paste output here
```

</details>

### PR DESCRIPTION
**Pull request by daixiang0**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #2385 #2323

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
Admission: support edge apps autonomy when offline
```

### CODE REVIEW

File: build/admission/deployment.yaml [hunk 1]
```diff
@@ -15,6 +15,17 @@ spec:
...
+        nodeAffinity:
+          requiredDuringSchedulingIgnoredDuringExecution:
+            nodeSelectorTerms:
+            - matchExpressions:
+              - key: node-role.kubernetes.io/master
+                operator: Exists
+      tolerations:
+      - key: node-role.kubernetes.io/master
+        operator: Exists
+        effect: NoSchedule
```
Comment (fisherxu) [inline build/admission/deployment.yaml:28]: We should not require webhook must to run on the master.

File: cloud/pkg/admissioncontroller/mutate_offlinemigration.go [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+}
+
+func generatePatch(tolerations []corev1.Toleration) []patchMapValue {
+	patch := []patchMapValue{{
+		Op:   "add",
+		Path: "/spec/template/spec/tolerations",
+		Value: []map[string]string{{
+			"key":      corev1.TaintNodeUnreachable,
+			"operator": "Exists",
+			"effect":   "NoExecute",
```
Comment (fisherxu) [inline cloud/pkg/admissioncontroller/mutate_offlinemigration.go:71]: Use const here.

File: cloud/pkg/admissioncontroller/mutate_offlinemigration.go [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+		}},
+	}}
+	if len(tolerations) > 0 {
+		for _, toleration := range tolerations {
+			if toleration.Key == corev1.TaintNodeUnreachable {
+				if toleration.Effect == "NoExecute" &&
+					toleration.Operator == "Exists" {
+					return nil
+				}
+				patch[0].Op = "replace"
```
Comment (fisherxu) [inline cloud/pkg/admissioncontroller/mutate_offlinemigration.go:81]: Now K8s will add toleration to pod as below:
```
  - effect: NoExecute
    key: node.kubernetes.io/unreachable
    operator: Exists
    tolerationSeconds: 300
```
Do we need to delete the deafultTimeout here?

File: build/admission/deployment.yaml (hunk 1 above)
Comment (daixiang0) [reply build/admission/deployment.yaml:28]: Run it at edge side has some problem, it needs direct connection to apiserver i think.

File: cloud/pkg/admissioncontroller/mutate_offlinemigration.go (hunk 3 above)
Comment (daixiang0) [reply cloud/pkg/admissioncontroller/mutate_offlinemigration.go:81]: I think so, or it would restart when timeout.

Comment (fisherxu) [reply cloud/pkg/admissioncontroller/mutate_offlinemigration.go:81]: So please fix it.

File: build/admission/deployment.yaml (hunk 1 above)
Comment (fisherxu) [reply build/admission/deployment.yaml:28]: Can also run on other cloud node

Comment (daixiang0) [reply build/admission/deployment.yaml:28]: I see, in fact we need add tolerations to skip edge nodes here.

File: cloud/pkg/admissioncontroller/admission.go [hunk 4]
```diff
@@ -222,31 +223,49 @@ func (ac *AdmissionController) registerWebhooks(opt *options.AdmissionOptions, c
 		},
 	}
 
-	return registerValidateWebhook(ac.Client.AdmissionregistrationV1beta1().ValidatingWebhookConfigurations(),
-		[]admissionregistrationv1beta1.ValidatingWebhookConfiguration{deviceModelCRDWebhook})
-}
+	objectSelector, err := metav1.ParseToLabelSelector("app-role.kubernetes.io=edge")
```
Comment (fisherxu) [inline cloud/pkg/admissioncontroller/admission.go:226]: The label in edge app here should reflect the migration info, how about:
```suggestion
	objectSelector, err := metav1.ParseToLabelSelector("app-offline.kubeedge.io=autonomy")
```
cc @kevin-wangzefeng

File: build/admission/deployment.yaml [hunk 5]
```diff
@@ -15,6 +15,14 @@ spec:
...
         app: kubeedge-admission
     spec:
+      affinity:
+        nodeAffinity:
+          requiredDuringSchedulingIgnoredDuringExecution:
+            nodeSelectorTerms:
+            - matchExpressions:
+              - key: node-role.kubernetes.io/edge
+                operator: DoesNotExist
+        effect: NoSchedule
```
Comment (fisherxu) [inline build/admission/deployment.yaml:25]: Delete this line?

File: build/admission/deployment.yaml (hunk 1 above)
Comment (fisherxu) [reply build/admission/deployment.yaml:25]: Looks good.

File: cloud/pkg/admissioncontroller/admission.go [hunk 6]
```diff
@@ -222,31 +223,49 @@ func (ac *AdmissionController) registerWebhooks(opt *options.AdmissionOptions, c
...
+				ObjectSelector: objectSelector,
+				Rules: []admissionregistrationv1beta1.RuleWithOperations{{
+					Operations: []admissionregistrationv1beta1.OperationType{
+						admissionregistrationv1beta1.Create,
+						admissionregistrationv1beta1.Update,
+					},
+					Rule: admissionregistrationv1beta1.Rule{
+						APIGroups:   []string{"apps", "extensions", ""},
+						APIVersions: []string{"v1", "v1beta1"},
+						Resources:   []string{"deployments", "daemonsets", "statefulsets"},
```
Comment (kevin-wangzefeng) [inline cloud/pkg/admissioncontroller/admission.go:246]: should detect the labels on pods not workloads
```suggestion
						APIGroups:   []string{""},
						APIVersions: []string{"v1"},
						Resources:   []string{"pods"},
```

File: cloud/pkg/admissioncontroller/admission.go (hunk 4 above)
Comment (kevin-wangzefeng) [reply cloud/pkg/admissioncontroller/admission.go:226]: +1 to `app-offline.kubeedge.io=autonomy`
And please define `app-offline.kubeedge.io=autonomy` as a const.

File: cloud/pkg/admissioncontroller/mutate_offlinemigration.go [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+			return toAdmissionResponse(err)
+		}
+		payload = generatePatch(statefulset.Spec.Template.Spec.Tolerations)
+	case "DaemonSet":
+		var daemonset v1.DaemonSet
+		if err := json.Unmarshal(review.Request.Object.Raw, &daemonset); err != nil {
+			klog.Errorf("Could not unmarshal raw object: %v", err)
+			return toAdmissionResponse(err)
+		}
+		payload = generatePatch(daemonset.Spec.Template.Spec.Tolerations)
```
Comment (fisherxu) [inline cloud/pkg/admissioncontroller/mutate_offlinemigration.go:52]: Remove these lines

**Comment by fisherxu** [comment]
Please fix the ci issue

**Comment by fisherxu** [comment]
The dco check is still failed

**Pull request merged**